libkbfs/interfaces.go

```diff
@@ -1648,12 +1655,39 @@ type MutableBareRootMetadata interface {
 	// BareRootMetadata. This is necessary since newly-created
 	// BareRootMetadata objects don't have enough data to build a
 	// TlfHandle from until the first rekey.
-	FakeInitialRekey(h BareTlfHandle)
+	FakeInitialRekey(c Codec, h BareTlfHandle) (ExtraMetadata, error)
 	// Update initializes the given freshly-created BareRootMetadata object with
 	// the given TlfID and BareTlfHandle. Note that if the given ID/handle are private,
 	// rekeying must be done separately.
 	// Update implements the BareRootMetadata interface for BareRootMetadataV2.
 	Update(tlf TlfID, h BareTlfHandle) error
 	// Returns the TLF key bundles for this metadata at the given key generation.
-	GetTLFKeyBundles(keyGen KeyGen) (*TLFWriterKeyBundle, *TLFReaderKeyBundle, error)
+	GetTLFKeyBundles(keyGen KeyGen) (*TLFWriterKeyBundleV2, *TLFReaderKeyBundleV2, error)
+	// GetUserDeviceKeyInfoMaps returns the given user device key info maps for the given
+	// key generation.
+	GetUserDeviceKeyInfoMaps(keyGen KeyGen, extra ExtraMetadata) (
+		readers, writers UserDeviceKeyInfoMap, err error)
+	// fillInDevices ensures that every device for every writer and reader
+	// in the provided lists has complete TLF crypt key info, and uses the
+	// new ephemeral key pair to generate the info if it doesn't yet
+	// exist.
+	// Note "wkb" is expcted to be nil for post-v3 metadata and "wkb2" is expected to be
+	// nil for pre-v3 metadata.
+	fillInDevices(crypto Crypto,
+		wkb *TLFWriterKeyBundleV2, wkb2 *TLFWriterKeyBundleV3, rkb *TLFReaderKeyBundleV2,
+		wKeys map[keybase1.UID][]CryptPublicKey,
+		rKeys map[keybase1.UID][]CryptPublicKey, ePubKey TLFEphemeralPublicKey,
+		ePrivKey TLFEphemeralPrivateKey, tlfCryptKey TLFCryptKey) (serverKeyMap, error)
+}
+
+// KeyBundleCache is an interface to a key bundle cache for use with v3 metadata.
+type KeyBundleCache interface {
```

Comment:
Hmm I don't see this used anywhere. But does it need to be an interface, since it's V3 specific? And should the functions all be named V3 and return V3 types?